File: src/platforms/twitter.ts
import { createHmac, randomBytes } from "node:crypto";
import type { AppConfig } from "../config.js";
import type { ComposedPost, Lot, Publisher, PublishResult } from "../domain/types.js";
import { PublishError } from "./bluesky.js";

export class TwitterPublisher implements Publisher {
  readonly platform = "twitter" as const;
  constructor(private readonly config: NonNullable<AppConfig["twitter"]>) {}

  private authorization(method: string, url: string): string {
    const oauth: Record<string, string> = {
      oauth_consumer_key: this.config.appKey,
      oauth_nonce: randomBytes(18).toString("hex"),
      oauth_signature_method: "HMAC-SHA1",
      oauth_timestamp: String(Math.floor(Date.now() / 1000)),
      oauth_token: this.config.accessToken,
      oauth_version: "1.0"
    };
    const encode = (value: string): string => encodeURIComponent(value).replace(/[!'()*]/g, (character) => `%${character.charCodeAt(0).toString(16).toUpperCase()}`);
    const parameters = Object.entries(oauth).sort(([left], [right]) => left.localeCompare(right))
      .map(([key, value]) => `${encode(key)}=${encode(value)}`).join("&");
    const base = `${method}&${encode(url)}&${encode(parameters)}`;
    const signingKey = `${encode(this.config.appSecret)}&${encode(this.config.accessSecret)}`;
    oauth.oauth_signature = createHmac("sha1", signingKey).update(base).digest("base64");
    return `OAuth ${Object.entries(oauth).sort(([left], [right]) => left.localeCompare(right))
      .map(([key, value]) => `${encode(key)}="${encode(value)}"`).join(", ")}`;
  }

  async publish(_lot: Lot, post: ComposedPost, image: Uint8Array): Promise<PublishResult> {
    const uploadUrl = "https://upload.twitter.com/1.1/media/upload.json";
    const form = new FormData();
    const imageCopy = new Uint8Array(image.byteLength);
    imageCopy.set(image);
    form.set("media", new Blob([imageCopy.buffer], { type: "image/jpeg" }), "image.jpg");
    const uploadResponse = await fetch(uploadUrl, {
      method: "POST",
      headers: { Authorization: this.authorization("POST", uploadUrl) },
      body: form
    });
    if (!uploadResponse.ok) throw new PublishError(`Twitter media upload returned HTTP ${uploadResponse.status}`, false);
    const upload = await uploadResponse.json() as { media_id_string?: string };
    if (upload.media_id_string === undefined) throw new PublishError("Twitter media upload response had no media ID", false);
    const postUrl = "https://api.x.com/2/tweets";
    try {
      const response = await fetch(postUrl, {
        method: "POST",
        headers: { Authorization: this.authorization("POST", postUrl), "Content-Type": "application/json" },
        body: JSON.stringify({ text: post.text, media: { media_ids: [upload.media_id_string] } })
      });
      if (!response.ok) throw new Error(`HTTP ${response.status}`);
      const result = await response.json() as { data?: { id?: string } };
      if (result.data?.id === undefined) throw new Error("response had no post ID");
      return { ref: result.data.id };
    } catch (error) {
      throw new PublishError("Twitter post write failed with an uncertain outcome", true, { cause: error });
    }
  }
}

File: src/twitter.rs
use crate::{
    config::Config,
    domain::Post,
    http::{self, Http},
};
use anyhow::{Result, ensure};
use base64::Engine;
use hmac::{Hmac, Mac};
use serde_json::json;
use sha1::Sha1;
use uuid::Uuid;
fn encode(value: &str) -> String {
    let mut result = String::new();
    for b in value.bytes() {
        if b.is_ascii_alphanumeric() || b"-._~".contains(&b) {
            result.push(b as char);
        } else {
            use std::fmt::Write;
            let _ = write!(result, "%{b:02X}");
        }
    }
    result
}
pub fn authorization(config: &Config, url: &str, nonce: &str, timestamp: &str) -> Result<String> {
    let mut params = vec![
        (
            "oauth_consumer_key",
            config.required("TWITTER_CONSUMER_KEY")?.to_owned(),
        ),
        ("oauth_nonce", nonce.to_owned()),
        ("oauth_signature_method", "HMAC-SHA1".to_owned()),
        ("oauth_timestamp", timestamp.to_owned()),
        (
            "oauth_token",
            config.required("TWITTER_ACCESS_TOKEN")?.to_owned(),
        ),
        ("oauth_version", "1.0".to_owned()),
    ];
    params.sort_by_key(|(key, _)| *key);
    let parameters = params
        .iter()
        .map(|(k, v)| format!("{}={}", encode(k), encode(v)))
        .collect::<Vec<_>>()
        .join("&");
    let base = format!("POST&{}&{}", encode(url), encode(&parameters));
    let signing = format!(
        "{}&{}",
        encode(config.required("TWITTER_CONSUMER_SECRET")?),
        encode(config.required("TWITTER_ACCESS_TOKEN_SECRET")?)
    );
    let mut mac = Hmac::<Sha1>::new_from_slice(signing.as_bytes())?;
    mac.update(base.as_bytes());
    params.push((
        "oauth_signature",
        base64::engine::general_purpose::STANDARD.encode(mac.finalize().into_bytes()),
    ));
    params.sort_by_key(|(key, _)| *key);
    Ok(format!(
        "OAuth {}",
        params
            .iter()
            .map(|(k, v)| format!("{}=\"{}\"", encode(k), encode(v)))
            .collect::<Vec<_>>()
            .join(", ")
    ))
}
fn auth(config: &Config, url: &str) -> Result<String> {
    authorization(
        config,
        url,
        &Uuid::new_v4().simple().to_string(),
        &chrono::Utc::now().timestamp().to_string(),
    )
}
pub fn upload(config: &Config, http: &Http, image: &[u8]) -> Result<String> {
    let url = "https://upload.twitter.com/1.1/media/upload.json";
    let boundary = format!("everylot-{}", Uuid::new_v4().simple());
    let mut multipart=format!("--{boundary}\r\nContent-Disposition: form-data; name=\"media\"; filename=\"image.jpg\"\r\nContent-Type: image/jpeg\r\n\r\n").into_bytes();
    multipart.extend_from_slice(image);
    multipart.extend_from_slice(format!("\r\n--{boundary}--\r\n").as_bytes());
    let result = http::json(http.bytes(
        url,
        &auth(config, url)?,
        &format!("multipart/form-data; boundary={boundary}"),
        &multipart,
    )?)?;
    Ok(result["media_id_string"]
        .as_str()
        .filter(|v| !v.is_empty())
        .ok_or_else(|| anyhow::anyhow!("Twitter upload response missing media ID"))?
        .to_owned())
}
pub fn create(config: &Config, http: &Http, post: &Post, media: &str) -> Result<String> {
    let url = "https://api.x.com/2/tweets";
    let result = http::json(http.post(
        url,
        Some(&auth(config, url)?),
        &json!({"text":post.text,"media":{"media_ids":[media]}}),
    )?)?;
    let id = result["data"]["id"].as_str().unwrap_or_default();
    ensure!(!id.is_empty(), "Twitter create response missing ID");
    Ok(id.to_owned())
}
